src/custom-element.js

```diff
@@ -394,7 +405,14 @@ function createBaseAmpElementProto(win) {
 
     /** @private {!./base-element.BaseElement} */
     this.implementation_ = new Ctor(this);
-    this.implementation_.createdCallback();
+
+    /**
+     * An element always starts in a unupgraded state until it's added to DOM
+     * for the first time in which case it can be upgraded immediately or wait
+     * for script download or `upgradeCallback`.
+     * @private {!UpgradeState}
+     */
+    this.upgradeState_ = UpgradeState.NOT_UPGRADED;
```

Comment:
I did it mostly for tests. Currently, the upgrade doesn't work via `createdCallback` because the `upgradeCallback` normally needs attributes and when calling `document.createElement('amp-element')`, the attributes are not available. For this reason, most of tests that need this kind of upgrade currently do it via `innerHTML`. See for example: https://github.com/ampproject/amphtml/blob/master/extensions/amp-carousel/0.1/test/test-slidescroll.js#L42

So, I decided to defer it until attachment. In practice this shouldn't add any latency, but will improve tests.